functions/nfts/asset/[[index]].ts

```diff
@@ -0,0 +1,19 @@
+/* eslint-disable import/no-unused-modules */
+import { MetaTagInjector } from '../../components/assetInjector'
+import getAsset from '../../utils/getAsset'
+
+export const onRequest: PagesFunction = async ({ params, request, next }) => {
+  try {
+    const { index } = params
+    const collectionAddress = String(index[0])
+    const tokenId = String(index[1])
+    const data = await getAsset(collectionAddress, tokenId, request.url)
+    if (!data) {
+      return await next()
+    }
+    return new HTMLRewriter().on('head', new MetaTagInjector(data)).transform(await next())
```

Comment:
You can parallelize this, but you'll need to save a reference to `res` to return in case of rejections

```
const assetPromise = getAsset(collectionAddress, tokenId, request.url)
const resPromise = next()
try {
  const [asset, res] = await Promise.all([assetPromise, resPromise])
  ...
} catch (e) {
  return resPromise
}
```